### PR TITLE
Fix TS types, add TSDoc comments

### PR DESCRIPTION
**Pull request by jedmao**
Fixes https://github.com/plopjs/plop/issues/180

### CODE REVIEW

**Comment by amwmedia** [comment]
there must be something wrong with the GH integration. I checked CircleCI and everything is passing for this build 👍

File: index.d.ts [hunk 1]
```diff
@@ -5,103 +5,322 @@ import { GlobbyOptions } from 'globby';
...
   force: boolean;
-  data: object;
+  /**
+   *
+   * @default {}
+   */
+  data:
+    | object
+    | ((...args: any[]) => object)
+    | ((...args: any[]) => Promise<object>);
```
Comment (cspotcode) [inline index.d.ts]: Should this be changed to `((...args: any[]) => object | Promise<object>)`?

Comment (jedmao) [reply index.d.ts]: See 74c372283de75feb75e43849dbdc582d76ddea19

**Comment by cspotcode** [comment]
Now that TypeScript 3.7 is (almost) complete, we could theoretically move all these declarations and comments into the codebase itself, and run `tsc` to extract these declarations from the codebase.  A big new feature is the support for emitting declarations while using allowJs.

**Comment by jedmao** [comment]
Why not just convert the project into TypeScript? 😄

**Comment by cspotcode** [comment]
> Why not just convert the project into TypeScript? 😄

Exactly!  :)  Often developers are opposed to that, so I try to pick my battles.  But I agree that moving to TypeScript all at once is ideal.

**Comment by crutchcorn** [comment]
I wouldn't suggest moving things into the project. There's still massive limitations in TS 3.7's AllowJS method. Generics aren't well supported and neither are interfaces (oftentimes exported type objects are used instead - a detremenet to those looking to extend interfaces). 

Also, I think the question is "why convert the project to TS". I don't think it's necessary for everything to be TS even if I do like TS a lot myself. It can add added complexity with build tools, make specific typing knowledge now required, and more

**Comment by cspotcode** [comment]
Now that `.d.ts` are kept in the codebase, they need to be kept up-to-date with the API, which requires type knowledge.  `// @ts-nocheck` can be used to disable semantic failures, easing development.  I imagine that any PR changing the API can be written and tested without getting the types exactly right.  Then someone with type knowledge can review, correcting types as needed.  This isn't much different than the current workflow, where someone needs to update the `.d.ts` separately.

I think nowadays type objects can be extended by interfaces.  Generic types can be declared well enough via .js.

I imagine that interfaces and advanced generics would be declared in .d.ts or .ts files, then imported by the .js.

**Comment by jedmao** [comment]
It's not _necessary_ in the same way that it's not _necessary_ to put syrup on my 🥞, but if you want to provide TypeScript support for your library, the absolute best way to do it is if the library itself is written in TypeScript and the definitions are generated, hands down. Everything else is a half-assed approach that comes with syncronization issues and inconsistencies.

If the concerns are build tool complexity, that's a one time fix.

If it's knowledge, that same knowledge needs to be applied to the `.d.ts` file; otherwise, why have the file at all? Are you comfortable publishing changes to your API that aren't also reflected in the definitions?

The #1 thing I fight for is the developer experience and there's no better way to provide the best developer experience than first-class TypeScript support (i.e., the library itself is written in TS).

This library is small enough that the conversion wouldn't be so bad.

That said, none of this discussion should be holding back this PR from being merged.

**Comment by amwmedia** [comment]
released in `plop@2.5.3`

**Pull request merged**